src/infra/core/src/dependency_graph_service_inmem.rs

```diff
@@ -193,6 +194,100 @@ impl DependencyGraphService for DependencyGraphServiceInMemory {
             .await
     }
 
+    async fn get_all_upstream_dependencies(
+        &self,
+        dataset_ids: Vec<DatasetID>,
+    ) -> Result<DatasetIDStream, GetUpstreamDependenciesError> {
+        self.ensure_datasets_initially_scanned()
+            .await
+            .int_err()
+            .map_err(GetUpstreamDependenciesError::Internal)
+            .unwrap();
+
+        let state = self.state.lock().await;
+
+        let nodes_to_search = dataset_ids
+            .iter()
+            .map(|dataset_id| {
+                state
+                    .get_dataset_node(dataset_id)
+                    .map_err(GetUpstreamDependenciesError::DatasetNotFound)
+            })
+            .collect::<Result<Vec<_>, _>>()?;
+
+        let mut result = Vec::new();
+        let mut visited_indexes = HashSet::new();
+
+        let mut reversed_graph = state.datasets_graph.clone();
+
+        reversed_graph.reverse();
```

Comment:
Let's put a `// TODO: PERF: ...` comment here as this looks like a quite expensive operation.

Reply:
I already have custom solution for recursive upstrean bradth-first search. I will provide it here as alternative

Reply:
This rather looks as a wrong idea. The library allows traversing in any order.
Both DFS and BFS algorithms call a function called "neighbors".
By default this function is returning outgoing edges, but it's possible to switch to incoming edges.

When I mentioned graph reversing in a Slack discussion, I did not mean "go clone a graph each time and reverse it physically". I meant a `Reversed` adapter from the library, that does not copy anything:

```
        let reversed_graph = Reversed(&state.datasets_graph);
```

Note - it's taking a reference, thus not copying neither taking ownership.
The only note, to access `node_weight`, use `.0` suffix to extract the original graph from the adapter's tuple struct.